hadoop-hdds/docs/content/design/storage-class.md

```diff
@@ -19,10 +20,408 @@ author: Marton Ele
+
+The exact implementation of these features (depends from further design discussion) can be different, but examples shows that the abstraction can make easier to implement new features.
+
+## Storage-class as configuration unit
+
+This document focuses on configuring replication mechanism for storage classes while the keys can use from pre-defined storage-classes.
+
+It's also possible to extend these options and define different topology rules, disk placement rules (SSD, HDD, ...) on prefix/key level.
+
+Storage-class is nothing more just the classification of containers to define the behavior and a binding between keys and containers. Any container specific configuration can be added leter.
+
+## Random-read write
+
+NFS / Fuse file system might require to support Random read/write which can be tricky as the closed containers are immutable. In case of changing any one byte in a block, the whole block should be re-created with the new data. It can have a lot of overhead especially in case of many small writes.
+
+But write is cheap with Ratis/THREE containers. Similar to any `ChunkWrite` and `PutBlock` we can implement an `UpdateChunk` call which modifies the current content of the chunk AND replicates the change with the help of Ratis.
+
+Let's imagine that we solved the resiliency of Ratis pipelines: In case of any Ratis error we can ask other Datanode to join to the Ratis ring *instead of* closing it. I know that it can be hard to implement, but if it is solved, we have an easy solution for random read/write.
+
+If it works, we can define the following storage-class:
+
+ * **Initial container** type/parameters: Ratis/THREE
+ * **Transitions**: NO. They will remain "Open" forever
+ * **Second container type**: NONE
+
+This would help the random read/write problem, with some limitations: only a few containers should be managed in this storage-class. It's not suitable for really Big data, but can be a very powerful addition to provide NFS/Fuse backend for containers / git db / SQL db.
+
+
+## Async, container-level erasure coding
+
+To store cold data on less bits (without the 3x replication overhead) we can encode the data with Erasure Coding and store some additional bits to survive replica loss. In a streaming situation (streaming write) it can be tricky as we need enough chunks to do the Reed-Solomon magic. With containers we are in a better position. After the first transition of the Open containers we can do EC encoding and that time we have enough data to encode.
+
+There are multiple options to do EC, one most simplest (but somewhat limited) way is to encode Containers after the first transition:
+
+Storage class COLD:
+
+ * *First container type/parameters*: Ratis/THREE replicated containers
+ * *Transitions*: In case of any error or if the container is full, convert to closed containers
+ * *Second container type/parameters*: EC RS(6,2)
```

Comment:
This is the same category as before, the beginning of the section defines it's a possible way to move forward.

This part is written to the container level erasure coding (not for the striped), which may or may not be implemented in addtional to the striped level. (I don't think we need it, but there were some opinions that striped EC can have some performance problems for Hive, where multiple disk read is required for small files instead of one disk read which can be cached by the OS).

But back to your question: let's say we have both container level async AND striped EC. Storage class is nothing more just an additional tag on the containers to classify which replication mechanism sholud be used. With this label we can classify containers easily and choose the right containers for a specific client request.